envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py

```diff
@@ -15,59 +19,106 @@
 
 
 if ipykernel_available:
-    from ipykernel.iostream import IOPubThread
-    from ipykernel.kernelapp import IPKernelApp
+    import ipykernel.iostream
+    import ipykernel.ipkernel
+    import ipykernel.kernelapp
+    import ipykernel.zmqshell
+    import IPython.utils.io
+    import zmq
 
     from envisage.plugins.ipython_kernel.internal_ipkernel import (
         InternalIPKernel)
 
 
+# Machinery for catching logged messages in tests (both for testing
+# the existence and content of those messages, and for hiding the
+# logger output during the test run).
+
+class ListHandler(logging.Handler):
+    """
+    Simple logging handler that stores all records in a list.
+
+    Used by catch_logs below.
+    """
+    def __init__(self):
+        logging.Handler.__init__(self)
+        self.records = []
+
+    def emit(self, record):
+        self.records.append(record)
+
+
 @contextlib.contextmanager
-def temp_filename(filename):
-    tmpdir = tempfile.mkdtemp()
+def catch_logs(logger_name, level=logging.WARNING):
+    """
+    Temporarily redirect logs for the given logger to a list.
+
+    Also allows temporarily changing the level of that logger.
+    """
+    logger = logging.getLogger(logger_name)
+    handler = ListHandler()
+    old_propagate = logger.propagate
+    old_level = logger.level
+
+    logger.addHandler(handler)
+    logger.propagate = False
+    logger.setLevel(level)
     try:
-        yield os.path.join(tmpdir, filename)
+        yield handler.records
     finally:
-        shutil.rmtree(tmpdir)
+        logger.setLevel(old_level)
+        logger.propagate = old_propagate
+        logger.removeHandler(handler)
 
 
 @contextlib.contextmanager
-def redirect_stdout_to(filename):
-    """ Redirect stdout to a file, in a way that's immune to changes
-    to Python's sys.stdout.
+def capture_stream(stream, callback=None):
+    """
```

Comment:
same comment as `catch_logs` here

Reply:
And the same reply. :-)

`apptools` may possibly be the right place for this sort of thing.